Abstract:
The data is related with direct marketing campaigns (phone calls) of a Portuguese banking institution. The classification goal is to predict if the client will subscribe a term deposit (variable y).

Data Set Information:
The data is related with direct marketing campaigns of a Portuguese banking institution. The marketing campaigns were based on phone calls. Often, more than one contact to the same client was required, in order to access if the product (bank term deposit) would be ('yes') or not ('no') subscribed.

Attribute Information:

Bank client data:
Age (numeric)
Job : type of job (categorical: 'admin.', 'blue-collar', 'entrepreneur', 'housemaid', 'management', 'retired', 'self-employed', 'services', 'student', 'technician', 'unemployed', 'unknown')

Marital : marital status (categorical: 'divorced', 'married', 'single', 'unknown' ; note: 'divorced' means divorced or widowed)

Education (categorical: 'basic.4y', 'basic.6y', 'basic.9y', 'high.school', 'illiterate', 'professional.course', 'university.degree', 'unknown')

Default: has credit in default? (categorical: 'no', 'yes', 'unknown')

Housing: has housing loan? (categorical: 'no', 'yes', 'unknown')

Loan: has personal loan? (categorical: 'no', 'yes', 'unknown')

Related with the last contact of the current campaign:


Contact: contact communication type (categorical:
'cellular','telephone')

Month: last contact month of year (categorical: 'jan', 'feb', 'mar',
…, 'nov', 'dec')

Day_of_week: last contact day of the week (categorical:
'mon','tue','wed','thu','fri')

Duration: last contact duration, in seconds (numeric). Important
note: this attribute highly affects the output target (e.g., if
duration=0 then y='no'). Yet, the duration is not known before a call
is performed. Also, after the end of the call y is obviously known.
Thus, this input should only be included for benchmark purposes and
should be discarded if the intention is to have a realistic
predictive model.



Other attributes:


Campaign: number of contacts performed during this campaign and for
this client (numeric, includes last contact)

Pdays: number of days that passed by after the client was last
contacted from a previous campaign (numeric; 999 means client was not
previously contacted)

Previous: number of contacts performed before this campaign and for
this client (numeric)

Poutcome: outcome of the previous marketing campaign (categorical:
'failure','nonexistent','success')

Social and economic context attributes
Emp.var.rate: employment variation rate - quarterly indicator
(numeric)

Cons.price.idx: consumer price index - monthly indicator (numeric)

Cons.conf.idx: consumer confidence index - monthly indicator
(numeric)

Euribor3m: euribor 3 month rate - daily indicator (numeric)

Nr.employed: number of employees - quarterly indicator (numeric)


Output variable (desired target):


y - has the client subscribed a term deposit? (binary: 'yes', 'no')

In [17]:
#first import pandas and read csv
import pandas as pd

In [18]:
 # read CSV file
df = pd.read_csv('bank-additional-full.csv', sep=';') 

In [19]:
#lets do an inital check of the data, with head(), info() and describe()
df.head()

age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         0  nonexistent          1.1   
1   may         mon  ...         1    999         0  nonexistent          1.1   
2   may         mon  ...         1    999         0  nonexistent          1.1   
3   may         mon  ...         1    999         0  nonexistent          1.1   
4   may         mon  ...         1    999         0  nonexistent          1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          93.994          -36.4      4.857       5191.0  no  
1          93.994          -36.4      4.857       5191.0  no  
2          93.994          -36.4      4.857       5191.0  no  
3          93.994          -36.4      4.857       5191.0  no  
4          93.994          -36.4      4.857       5191.0  no  

[5 rows x 21 columns]

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null 

In [21]:
df.describe()

age      duration      campaign         pdays      previous  \
count  41188.00000  41188.000000  41188.000000  41188.000000  41188.000000   
mean      40.02406    258.285010      2.567593    962.475454      0.172963   
std       10.42125    259.279249      2.770014    186.910907      0.494901   
min       17.00000      0.000000      1.000000      0.000000      0.000000   
25%       32.00000    102.000000      1.000000    999.000000      0.000000   
50%       38.00000    180.000000      2.000000    999.000000      0.000000   
75%       47.00000    319.000000      3.000000    999.000000      0.000000   
max       98.00000   4918.000000     56.000000    999.000000      7.000000   

       emp.var.rate  cons.price.idx  cons.conf.idx     euribor3m   nr.employed  
count  41188.000000    41188.000000   41188.000000  41188.000000  41188.000000  
mean       0.081886       93.575664     -40.502600      3.621291   5167.035911  
std        1.570960        0.578840       4.628198      1.734447     72.251528  
min       -3.400000       92.201000     -50.800000      0.634000   4963.600000  
25%       -1.800000       93.075000     -42.700000      1.344000   5099.100000  
50%        1.100000       93.749000     -41.800000      4.857000   5191.000000  
75%        1.400000       93.994000     -36.400000      4.961000   5228.100000  
max        1.400000       94.767000     -26.900000      5.045000   5228.100000

Notes:


1. age
Mean age is ~40 years, with a range from 17 to 98.

The distribution seems reasonable for adult clients/customers.

Std dev of ~10 means typical variation around 40 is ±10 years.




2. duration (likely call duration or event length in seconds)
Mean is ~258 seconds (~4 minutes 18 seconds).

Huge range from 0 to 4918 seconds (~82 minutes).

Median is 180 seconds, showing some longer calls/events pull the mean up.

Could be right-skewed with a few very long calls.

3. campaign (number of contacts performed during this campaign)
Mean about 2.57 contacts.

Max 56, which seems like an outlier (very high number of contacts).

Median = 2, with 75% at 3 or less, so most clients had few contacts.

4. pdays (days passed since last contact)
Mean around 962 days, but values cluster a lot at 999 (likely a placeholder meaning “not contacted before”).

25%, 50%, 75% all at 999, meaning most clients were not contacted before this campaign.

The min is 0, meaning some recent contacts.

5. previous (number of contacts performed before this campaign)
Mean 0.17, median 0, max 7.

Most clients had no prior contact, with a few having a handful.

6. emp.var.rate (employment variation rate)
Mean close to zero at 0.08.

Range from -3.4 to 1.4.

Std dev ~1.57, indicates some volatility in employment stats during the data collection period.

7. cons.price.idx (consumer price index)
Mean ~93.58 with a narrow range (~92.2 to 94.77).

Std dev low at 0.58, so price index fairly stable.

8. cons.conf.idx (consumer confidence index)
Mean ~ -40.5 (negative values).

Wide spread from -50.8 to -26.9.

Std dev of 4.6 indicates some variation in consumer confidence.

9. euribor3m (3-month Euribor interest rate)
Mean ~3.62%, range from 0.63 to 5.05.

Std dev ~1.73 shows moderate variability.

10. nr.employed (number of employees)
Mean ~5167, range from 4963.6 to 5228.1.

Std dev ~72 shows small variability, likely reflecting macroeconomic changes.

Summary observations:
Some columns like pdays and campaign show clear signs of placeholder or outlier values (999 for pdays, 56 contacts max).

Most numeric columns look like macroeconomic indicators with limited ranges, reflecting stable economic conditions with small fluctuations.

Duration and campaign data are skewed and may require transformation or special treatment in modeling.

In [28]:
#Identify outliers
#I'm going to use the IQR method to identify outliers in the dataframe
def identify_outliers_iqr(df, column):
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    outliers = df[(df[column] < lower_bound) | (df[column] > upper_bound)]
    return outliers, lower_bound, upper_bound

# Apply the function to duration, campaign, and pdays
outliers_duration, dur_low, dur_high = identify_outliers_iqr(df, 'duration')   
outliers_campaign, camp_low, camp_high = identify_outliers_iqr(df, 'campaign')
outliers_pdays, pdays_low, pdays_high = identify_outliers_iqr(df, 'pdays')

# Print summary
print(f"Duration outliers: {len(outliers_duration)} (above {dur_high:.2f})")
print(f"Campaign outliers: {len(outliers_campaign)} (above {camp_high:.2f})")
print(f"Pdays outliers: {len(outliers_pdays)} (below {pdays_low:.2f} or above {pdays_high:.2f})")

Duration outliers: 2963 (above 644.50)
Campaign outliers: 2406 (above 6.00)
Pdays outliers: 1515 (below 999.00 or above 999.00)



🟡 Duration outliers: 2963 (above 644.50)
You have 2,963 outliers where call duration > 644.5 seconds (~10 minutes 45 seconds).

Since the median duration is 180 sec (3 minutes), this confirms the data is right-skewed — most calls are short, but a few are extremely long.

These long durations could be real (e.g., high-interest customers), but may also distort modeling.

👉 Options:

Keep them if they’re meaningful for predicting success.

Log-transform or cap for modeling to reduce their influence.

Create a binary flag: is_long_call = duration > 644.5.

🟡 Campaign outliers: 2406 (above 6.00)
2,406 clients were contacted more than 6 times.

Since 75% of clients were contacted 3 or fewer times, this is a very aggressive follow-up for those 2406.

Could be signal: persistence might correlate with success or failure.

👉 Options:

Cap at 6, or log-transform if needed.

Create a flag: high_contact = campaign > 6.

Explore correlation between # of contacts and positive outcomes.

🔴 Pdays outliers: 1515 (below 999.00 or above 999.00)
This one is special:

The output says that 1,515 values are "below 999 or above 999".

That means 999 is the only value NOT counted as an outlier.

In your data, 999 is used as a placeholder, meaning “not previously contacted.”

The actual values (like 0–998) are real days since last contact.

👉 So, this is reversed:

999 is the oddball, not the others.

You should treat 999 as a sentinel/missing value — not real time.